### PR TITLE
[BROWSEUI] Implement auto-completion

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implement `IAutoComplete` to realize input auto completion.

JIRA issue: [CORE-9281](https://jira.reactos.org/browse/CORE-9281), [CORE-1419](https://jira.reactos.org/browse/CORE-1419)

## Proposed changes

- The main features of this PR are Auto-Suggest and Auto-Append.
- Auto-Suggest shows a list near the textbox (an EDIT control) when the user has typed partial pathname into the textbox.
- Auto-Append appends complement text into the textbox to complete the pathname with selected status.
- The list of AutoSuggest is a top-level window whose window class is `"Auto-Suggest Dropdown"`. We call it "the drop-down window".
- The drop-down window contains three controls: a listview, a scrollbar and a sizebox.
- If the list is short, then the scrollbar might be invisible.
- The drop-down window watches the input into the textbox. If the textbox changed, then the window updates the list.
- The sizebox control enables the user to resize the drop-down window.

## TODO

- [x] AutoSuggest works 
- [x] ScrollBar works
- [x] ListView works
- [x] SizeBox works
- [x] Keyboard works
- [x] Mouse works
- [x] Focus OK
- [x] AutoAppend works 
- [x] Ctrl+Back works
- [x] Enter key works

### CODE REVIEW

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_06_03_2021_12_59_12
ListView won't be drawn.

**Comment by katahiromz** [comment]
gotcha
ListView is drawn. How about scrollbar?

**Comment by katahiromz** [comment]
scrollbar
Now scrollbar works.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_07_03_2021_17_06_25
size-box works.

**Comment by katahiromz** [comment]
Z Order is broken. It's not my fault.
https://jira.reactos.org/browse/CORE-6651
VirtualBox_ReactOS_08_03_2021_16_51_06

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_08_03_2021_17_06_08

**Comment by katahiromz** [comment]
BEFORE:
```txt
# browseui_winetest.exe autocomplete
autocomplete: 65 tests executed (0 marked as todo, 0 failures), 1 skipped.
# shell32_winetest.exe autocomplete
autocomplete: 215 tests executed (0 marked as todo, 108 failures), 0 skipped.
# browseui_apitest.exe ACListISF
ACListISF: 528 tests executed (0 marked as todo, 0 failures), 0 skipped.
# browseui_apitest.exe IACLCustomMRU
IACLCustomMRU: 331 tests executed (0 marked as todo, 0 failures), 0 skipped.
# browseui_apitest.exe IAutoComplete
IAutoComplete: 739 tests executed (0 marked as todo, 301 failures), 1 skipped.
```

AFTER:
```txt
# browseui_winetest.exe autocomplete
autocomplete: 65 tests executed (0 marked as todo, 0 failures), 1 skipped.
# shell32_winetest.exe autocomplete
autocomplete: 214 tests executed (0 marked as todo, 48 failures), 0 skipped.
# browseui_apitest.exe ACListISF 
ACListISF: 528 tests executed (0 marked as todo, 0 failures), 0 skipped.
# browseui_apitest.exe IACLCustomMRU
IACLCustomMRU: 331 tests executed (0 marked as todo, 0 failures), 0 skipped.
# browseui_apitest.exe IAutoComplete
IAutoComplete: 66275 tests executed (0 marked as todo, 56 failures), 0 skipped.
```

File: dll/win32/browseui/CAutoComplete.cpp [hunk 1]
```diff
@@ -25,664 +25,2099 @@
...
+LRESULT CACEditCtrl::OnSetText(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
 {
-    INT iStart, iEnd;
-    iStart = iEnd = 0;
-    SendMessageW(hwndEdit, EM_GETSEL, (WPARAM)&iStart, (LPARAM)&iEnd);
+    TRACE("CACEditCtrl::OnSetText(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    if (!m_pDropDown->m_bInSetText)
+        m_pDropDown->HideDropDown(); // it's mechanical WM_SETTEXT
+    return DefWindowProcW(uMsg, wParam, lParam); // do default
```
Comment (yagoulas) [inline dll/win32/browseui/CAutoComplete.cpp:414]: I think that if you set bHandled to false the caller will automatically call DefWindowProcW

Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: OK.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 2]
```diff
@@ -25,664 +25,2124 @@
...
+// WM_CUT @implemented
+// An application sends this message to an edit control to cut the current selection.
+LRESULT CACEditCtrl::OnCut(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnCut(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // do default
+    m_pDropDown->OnEditUpdate(TRUE);
+    return ret;
+}
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:296]: This function seems to be identical to the OnClear() one. If they are really meant to be identical, maybe you could group them together? (and both WM_CUT and WM_CLEAR use the same one)

File: dll/win32/browseui/CAutoComplete.cpp [hunk 3]
```diff
@@ -25,664 +25,2124 @@
...
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // get default
+
+    if (m_pDropDown)
+    {
+        // some special keys need default processing. we handle them here
+        switch (wParam)
         {
-            TRACE("HeapAlloc Failed when trying to alloca %d bytes\n", (wcslen(pwszQuickComplete) + 1) * sizeof(WCHAR));
-            return S_FALSE;
+            case VK_RETURN: // [Enter] key
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:336]: I think everyone knows what these VK_xxx keys refer to.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 4]
```diff
@@ -25,664 +25,2124 @@
...
+    return 0;
 }
 
-/**************************************************************************
- *  IAutoComplete_fnSetOptions
- */
-HRESULT WINAPI CAutoComplete::SetOptions(DWORD dwFlag)
+// WM_PASTE @implemented
+// An application sends this message to an edit control to paste at the current caret position.
+LRESULT CACEditCtrl::OnPaste(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:391]: Again this function looks identical to the ones for WM_CUT and WM_CLEAR.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 5]
```diff
@@ -25,664 +25,2124 @@
...
 
-    CComHeapPtr<WCHAR> pszText;
-    if (!pszText.Allocate(cchText + 1))
-        return;
+HWND CACListView::Create(HWND hwndParent)
+{
+    ATLASSERT(m_hWnd == NULL);
+    DWORD dwStyle = WS_CHILD | /*WS_VISIBLE |*/ WS_CLIPSIBLINGS | LVS_NOCOLUMNHEADER |
+                    LVS_OWNERDATA | LVS_OWNERDRAWFIXED | LVS_SINGLESEL | LVS_REPORT;
+    HWND hWnd = ::CreateWindowExW(0, GetWndClassName(), L"Internet Explorer", dwStyle,
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:435]: Internet Explorer?! Windows really does that?

File: dll/win32/browseui/CAutoComplete.cpp [hunk 6]
```diff
@@ -25,664 +25,2124 @@
...
+{
+    TRACE("CACListView::OnMButtonDown(%p)\n", this);
+    return 0; // eat
+}
 
-        case WM_KEYUP:
-        {
-            GetWindowTextW(hwnd, (LPWSTR)hwndText, 255);
+// WM_MBUTTONUP @implemented
+LRESULT CACListView::OnMButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:553]: These three last mouse handlers (OnLButtonUp, OnMButtonDown, OnMButtonUp) seem to do the same thing; they can be grouped together.
Also I'm not sure you need to write handlers for the middle button (not wheel). (Actually, what is the default behaviour of any combobox when the middle button is pressed?)

File: dll/win32/browseui/CAutoComplete.cpp [hunk 7]
```diff
@@ -25,664 +25,2124 @@
...
+// WM_RBUTTONDOWN @implemented
+LRESULT CACListView::OnRButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACListView::OnRButtonDown(%p)\n", this);
+    return 0; // eat
+}
 
-            HeapFree(GetProcessHeap(), 0, pThis->m_txtbackup);
+// WM_RBUTTONUP @implemented
+LRESULT CACListView::OnRButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:591]: The same comment as for the middle button may apply as well for the right button.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 8]
```diff
@@ -25,664 +25,2124 @@
...
+    if (strCommon.IsEmpty() || strCommon.GetLength() <= strText.GetLength())
+        return; // no suggestion
+
+    // append suggestion
+    INT cchAppend = strCommon.GetLength() - strText.GetLength();
+    CStringW strAppend = strCommon.Right(cchAppend);
+    strText += strAppend;
+    SetEditText(strText);
+
+    // select the appended
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:900]:
```suggestion
    // select the appended suggestion
```
?

File: dll/win32/browseui/CAutoComplete.cpp [hunk 9]
```diff
@@ -25,664 +25,2124 @@
...
+    INT ich0 = strText.GetLength();
+    INT ich1 = ich0 + cchAppend;
+    SetEditSel(ich0, ich1);
+}
+
+// go back a word ([Ctrl]+[Backspace])
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:906]: I always wondered: Is this functionality an actual default one present in all edit text controls? (it seems to appear so in Windows, at least 10)

File: dll/win32/browseui/CAutoComplete.cpp [hunk 10]
```diff
@@ -25,664 +25,2124 @@
...
+BOOL CAutoComplete::OnEditKeyDown(WPARAM wParam, LPARAM lParam)
+{
+    TRACE("CAutoComplete::OnEditKeyDown(%p, %p)\n", this, wParam);
+
+    UINT vk = (UINT)wParam; // virtual key
+    switch (vk)
+    {
+        case VK_HOME: case VK_END: // [Home]/[End] key
+        case VK_UP: case VK_DOWN: // [Arrow Up]/[Arrow Down] key
+        case VK_PRIOR: case VK_NEXT: // [PageUp]/[PageDown] key
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:954]: No need for the comments that describe the keys, since this is well-known from the VK_ defines names.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 11]
```diff
@@ -25,664 +25,2124 @@
...
+    // load quick completion info
+    LoadQuickComplete(pwszRegKeyPath, pwszQuickComplete);
+
+    // any combobox for m_hwndEdit?
+    m_hwndCombo = NULL;
+    HWND hwndParent = ::GetParent(m_hwndEdit);
+    WCHAR szClass[16];
+    if (::GetClassNameW(hwndParent, szClass, _countof(szClass)))
+    {
+        if (::StrCmpIW(szClass, L"COMBOBOX") == 0 ||
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:1229]:
```suggestion
        if (::StrCmpIW(szClass, WC_COMBOBOXW) == 0 ||
```

File: dll/win32/browseui/CAutoComplete.cpp [hunk 12]
```diff
@@ -25,664 +25,2124 @@
...
+    // create the children
+    m_hwndList.Create(m_hWnd);
+    if (!m_hwndList)
+        return -1; // failure
+    m_hwndSizeBox.Create(m_hWnd);
+    if (!m_hwndSizeBox)
+        return -1; // failure
+    m_hwndScrollBar.Create(m_hWnd);
+    if (!m_hwndScrollBar)
+        return -1; // failure
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:1708]: Why is this needed to create all this? Don't they come automatically when using the combobox control? (and then, only retrieving the corresponding existing controls when they show up?)

File: dll/win32/browseui/CAutoComplete.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: United.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: United.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: United.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: Deleted comments about `VK_*`.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: United.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 9 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:861]: No. Notepad of Win2k3 generates a tofu block on Ctrl+BackSpace.

VirtualBox_Windows Server 2003_09_03_2021_12_30_35

File: dll/win32/browseui/CAutoComplete.cpp (hunk 10 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: Deleted.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 11 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: OK.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 8 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp]: OK.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:411]: I can confirm it by WinHier. This is the evidence:
evidence

File: dll/win32/browseui/CAutoComplete.cpp (hunk 12 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:1662]: WinHier says these controls should be created.

**Comment by katahiromz** [comment]
Ready to go!

File: dll/win32/browseui/CAutoComplete.cpp (hunk 9 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:851]: I can observe the window procedure change of the EDIT control in auto-completion textboxes.

**Comment by HBelusca** [comment]
I'm adding here things to think about for this PR (will be edited as things come out):

- Theme support (does the autocompleter list display ok/ "as expected" when themes are enabled? --> See SetWindowTheme() );

File: dll/win32/browseui/CAutoComplete.cpp [hunk 13]
```diff
@@ -20,669 +20,2074 @@
...
 }
 
-/*
-  Window procedure for autocompletion
- */
-LRESULT APIENTRY CAutoComplete::ACEditSubclassProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+// get the text of an item
+CStringW CACListView::GetItemText(INT iItem)
 {
-    CAutoComplete *pThis = static_cast<CAutoComplete *>(GetPropW(hwnd, autocomplete_propertyW));
```
Comment (HBelusca) [inline dll/win32/browseui/CAutoComplete.cpp:329]: Windows has this SetPropW / GetPropW with prop title "CAutoComplete_This" name for keeping the context.

Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:329]: However it's no use because ATL can associate HWND and pointer.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 12 above)
Comment (HBelusca) [reply dll/win32/browseui/CAutoComplete.cpp:1652]: Yes indeed, I've checked and can confirm, it's actually a hidden listview with scrollbar + resizing, that gets shown whenever autocompletion is needed, and just hidden when it's not (and created / destroyed, when an editable window that needs autocompletion is created/destroyed).

File: dll/win32/browseui/CAutoComplete.cpp (hunk 13 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:329]: GetProp/SetProp is slow.
